src/scene/morph.js

```diff
@@ -29,19 +49,162 @@ Object.assign(pc, function () {
+                    ids.push(0);
+                }
+            }
+
+            // max texture size: vertexBufferIds is stored in float32 format, giving us 2^24 range, so can address 4096 texture at maximum
+            // TODO: on webgl2 we could store this in uint32 format and remove this limit
+            var maxTextureSize = Math.min(this.device.maxTextureSize, 4096);
+
+            // texture size for freeIndex pixels - roughly square
+            var morphTextureWidth = Math.ceil(Math.sqrt(freeIndex));
+            morphTextureWidth = Math.min(morphTextureWidth, maxTextureSize);
+            var morphTextureHeight = Math.ceil(freeIndex / morphTextureWidth);
+
+            // if data cannot fit into max size texture, fail this set up
+            if (morphTextureHeight > maxTextureSize) {
+                return false;
+            }
+
+            this.morphTextureWidth = morphTextureWidth;
+            this.morphTextureHeight = morphTextureHeight;
+
+            // texture format based vars
+            var halfFloat = false;
+            var numComponents = 3;  // RGB32 is used
+            var float2Half = pc.math.float2Half;
+            if (this._textureFormat === pc.Morph.FORMAT_HALF_FLOAT) {
+                halfFloat = true;
+                numComponents = 4;  // RGBA16 is used, RGB16 does not work
+            }
+
+            // build texture for each delta array, all textures are the same size
+            var arraySize = this.morphTextureWidth * this.morphTextureHeight * numComponents;
+            var packedDeltas = halfFloat ? new Uint16Array(arraySize) : new Float32Array(arraySize);
+            for (i = 0; i < deltaArrays.length; i++) {
+                data = deltaArrays[i];
+
+                // copy full arrays into sparse arrays and convert format (skip 0th pixel - used by non-morphed vertices)
+                for (v = 0; v < usedDataIndices.length; v++) {
+                    var index = usedDataIndices[v];
+
+                    if (halfFloat) {
+                        packedDeltas[v * numComponents + numComponents] = float2Half(data[index * 3]);
+                        packedDeltas[v * numComponents + numComponents + 1] = float2Half(data[index * 3 + 1]);
+                        packedDeltas[v * numComponents + numComponents + 2] = float2Half(data[index * 3 + 2]);
+                    } else {
+                        packedDeltas[v * numComponents + numComponents] = data[index * 3];
+                        packedDeltas[v * numComponents + numComponents + 1] = data[index * 3 + 1];
+                        packedDeltas[v * numComponents + numComponents + 2] = data[index * 3 + 2];
+                    }
+                }
+
+                // create texture and assign it to target
+                target = deltaInfos[i].target;
+                var format = this._textureFormat === pc.Morph.FORMAT_FLOAT ? pc.PIXELFORMAT_RGB32F : pc.PIXELFORMAT_RGBA16F;
```

Comment:
Does this texture need to be color-renderable? RGB32F isn't per:
https://www.khronos.org/registry/OpenGL/extensions/EXT/EXT_color_buffer_float.txt

but it sounds like the code above already takes this into account, so maybe these textures are only sampled, not rendered into.

Reply:
yes, these are only the source texture for reading, either RGB32F, or RGBA16F (RBG16F was failing for me unless I created it power of 2)

Render target texture is created here:

https://github.com/playcanvas/engine/blob/2dce02f7a89d9fde2262a2b886865995806a5cd6/src/scene/morph-instance.js#L44-L46